### PR TITLE
`Quiz`: Implement quiz participation

### PR DESCRIPTION
**Pull request by VSchlattinger**
## Summary
Adds the initial student-facing quiz participation workflow. Quiz exercises in the course exercise list now open a quiz participation route where students can answer multiple-choice questions, see score, remaining time, and last-submission information, and submit answers while the quiz is still active.

## Motivation and Context
This enables students to participate in quiz exercises from the course page instead of only creating or managing quiz content. No issue is linked to this PR.

Original note from the PR: The Submit button behavior differs from what was discussed in the team call, and the socket connection is still missing.

## Description
- Adds the AngularJS quiz participation route, controller, and template under `src/main/webapp/app/quiz/participate/`, wires them in `index.html`, and updates quiz styles, labels, and English/German translations.
- Updates the course exercise list so quiz exercises navigate to `/quiz/{id}`, use quiz-specific start/resume text and status states, and rely on the backend `isVisibleToStudents` flag for release visibility.
- Extends the multiple-choice participation component so selected answer options are managed by the parent quiz controller, compared by answer option id, and displayed together with the question score.
- Adds student-safe quiz exercise retrieval via `/api/quiz-exercises/{id}/for-student`, filtering explanations and correct-answer markers from the response.
- Changes course exercise retrieval to filter by course permissions and student visibility on the backend, and removes quiz questions from exercise-list payloads.
- Adds quiz submission support with `/courses/{courseId}/exercises/{exerciseId}/submissions/my-latest` to initialize or retrieve the current participation/result/submission and `PUT /api/quiz-submissions` to save submitted answers and update the result completion date before the quiz ends.
- Extends quiz submission and submitted-answer persistence with a submitted-answer-to-question relation, JSON subtype metadata for submitted answer types, cascading eager quiz submitted answers, a transient submission date, repository helpers, Liquibase migration, and updated tests for the changed resource constructor.

## Steps for Testing
Prerequisites:
- A course with a planned quiz exercise containing multiple-choice questions.
- One instructor or teaching assistant and one student enrolled in the course.

1. Log in as the student and open the course exercise list after the quiz is visible to students.
2. Verify that the quiz row shows quiz-specific status text and Start Quiz or Resume Quiz, then start the quiz.
3. Verify that `/quiz/{exerciseId}` opens and shows the course title, quiz title, questions, question scores, total score, remaining time, and last-submission information.
4. Select and deselect multiple-choice answer options, submit the quiz, and verify that the last-submission display updates.
5. Reload or resume the quiz and verify that the previous selected answer options are loaded from the latest submission.
6. Verify that student-facing quiz responses do not expose correct-answer markers or explanations in the network payload.
7. Let the quiz time expire or use a quiz whose due date is in the past, then verify that submitting is disabled or rejected.
8. As instructor or teaching assistant, verify that the exercise list still shows unreleased exercises as not released, while students only receive exercises visible to them.

## Checklist
- [x] I have read and followed the [coding guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/)
- [x] PR title follows the Artemis naming convention.
- [x] PR description matches the actual diff.
- [ ] Manual quiz participation flow has been tested on a suitable test setup.
- [ ] Relevant automated tests have been run.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/exerciseapp/domain/SubmittedAnswer.java [hunk 1]
```diff
@@ -19,6 +22,11 @@
 )
 @DiscriminatorValue(value="S")
 @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/domain/SubmittedAnswer.java]: Please document the following lines properly so that other developers know why you are using these annotations

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizSubmissionResource.java [hunk 2]
```diff
@@ -27,8 +35,51 @@
...
+    public QuizSubmissionResource(QuizSubmissionRepository quizSubmissionRepository, QuizExerciseRepository quizExerciseRepository, ResultRepository resultRepository, ParticipationService participationService) {
         this.quizSubmissionRepository = quizSubmissionRepository;
+        this.quizExerciseRepository = quizExerciseRepository;
+        this.resultRepository = resultRepository;
+        this.participationService = participationService;
+    }
+
+    /**
+     * GET  /courses/{courseId}/exercises/{exerciseId}/submissions/my-latest : Get the latest quizSubmission for the given course.
+     *
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizSubmissionResource.java]: I don't really get what this call is used for. Is it that the user can get an "open" submission, in case he has lost his connection and reconnects?
Please document it properly.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizSubmissionResource.java [hunk 3]
```diff
@@ -61,16 +114,43 @@ public QuizSubmissionResource(QuizSubmissionRepository quizSubmissionRepository)
...
      */
     @PutMapping("/quiz-submissions")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'ADMIN')")
     @Timed
     public ResponseEntity<QuizSubmission> updateQuizSubmission(@RequestBody QuizSubmission quizSubmission) throws URISyntaxException {
         log.debug("REST request to update QuizSubmission : {}", quizSubmission);
+
+        //TODO: Valentin: check if submission belongs to user
+
+        // recreate pointers back to submission in each submitted answer
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizSubmissionResource.java]: why is this necessary? please add a meaningful comment

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizSubmissionResource.java [hunk 4]
```diff
@@ -61,16 +114,43 @@ public QuizSubmissionResource(QuizSubmissionRepository quizSubmissionRepository)
...
         if (quizSubmission.getId() == null) {
             return createQuizSubmission(quizSubmission);
         }
-        QuizSubmission result = quizSubmissionRepository.save(quizSubmission);
-        return ResponseEntity.ok()
-            .headers(HeaderUtil.createEntityUpdateAlert(ENTITY_NAME, quizSubmission.getId().toString()))
-            .body(result);
+        quizSubmission = quizSubmissionRepository.save(quizSubmission);
+
+        // update corresponding result
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizSubmissionResource.java]: please add more comments what happens here and why it is necessary

File: src/main/webapp/app/courses/exercises/exercise-list.component.js [hunk 5]
```diff
@@ -113,7 +113,15 @@
             if(angular.equals({}, exercise.participation)) {
                 return "uninitialized";
             } else if(exercise.participation.initializationState === "INITIALIZED") {
-                return "initialized";
+                if (exercise.type === "quiz") {
```
Comment (krusche) [inline src/main/webapp/app/courses/exercises/exercise-list.component.js]: please check if exercise.type is available, as below

File: src/main/webapp/app/courses/exercises/exercise-list.html [hunk 6]
```diff
@@ -45,6 +47,13 @@
                 <i class="fa fa-circle-o-notch fa-spin" ng-show="$ctrl.loading[exercise.id.toString()]"></i>
             </button>
             <button class="btn btn-primary btn-sm btn-block"
+                    ng-switch-when="active"
+                    ng-click="$ctrl.start(exercise)"
+                    ng-disabled="$ctrl.loading[exercise.id.toString()]">
+                <span ng-hide="$ctrl.loading[exercise.id.toString()]"><i class="fa fa-play-circle fa-fw"></i>&nbsp;Resume exercise</span>
```
Comment (krusche) [inline src/main/webapp/app/courses/exercises/exercise-list.html]: Should this not be "Resume quiz"?

File: src/main/webapp/app/entities/submitted-answer/submitted-answer-dialog.html [hunk 7]
```diff
@@ -15,7 +15,13 @@ <h4 class="modal-title" id="mySubmittedAnswerLabel" data-translate="artemisApp.s
         </div>
 
         <div class="form-group">
-            <label data-translate="artemisApp.submittedAnswer.submission" for="field_submission">Submission</label>
+            <label data-translate="arTeMiSApp.submittedAnswer.question" for="field_question">Question</label>
```
Comment (krusche) [inline src/main/webapp/app/entities/submitted-answer/submitted-answer-dialog.html]: I think it always has to be "artemisApp" instead of "arTeMiSApp". Please search and replace all occurrences!

File: src/main/webapp/app/quiz/participate/multiple-choice-question/multiple-choice-question.component.js [hunk 8]
```diff
@@ -19,19 +19,26 @@ function MultipleChoiceQuestionController($translate, $translatePartialLoader, $
...
     };
 
-    vm.selectedAnswerOptions = [];
     vm.toggleSelection = toggleSelection;
 
     function toggleSelection(answerOption) {
-        if (vm.selectedAnswerOptions.indexOf(answerOption) !== -1) {
+        if (isAnswerOptionSelected(answerOption)) {
             vm.selectedAnswerOptions = vm.selectedAnswerOptions.filter(function(ao) {
-                return ao !== answerOption;
```
Comment (krusche) [inline src/main/webapp/app/quiz/participate/multiple-choice-question/multiple-choice-question.component.js]: Don't use the variable name 'ao'

File: src/main/webapp/app/quiz/participate/quiz.controller.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+        // update displayed times in UI regularly
+        $interval(function () {
+            // update remaining time
+            if (vm.quizExercise && vm.quizExercise.adjustedDueDate) {
+                var endDate = vm.quizExercise.adjustedDueDate;
+                if (endDate.isAfter(moment())) {
+                    vm.remainingTimeSeconds = endDate.diff(moment(), "seconds");
+                    if (vm.remainingTimeSeconds > 90) {
+                        vm.remainingTime = Math.ceil(vm.remainingTimeSeconds / 60) + " minutes";
```
Comment (krusche) [inline src/main/webapp/app/quiz/participate/quiz.controller.js]: Why do you not show the seconds earlier? You could at least show e.g. "1 min 30 sec" if it is below 3 minutes. I would also abbreviate minutes to "min" and seconds to "sec" or even "s"

File: src/main/webapp/app/quiz/participate/quiz.controller.js [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                    } else if (vm.remainingTimeSeconds > 4) {
+                        vm.remainingTime = vm.remainingTimeSeconds + " seconds";
+                    } else {
+                        vm.remainingTime = "< 5 seconds";
```
Comment (krusche) [inline src/main/webapp/app/quiz/participate/quiz.controller.js]: Why do you not display the last seconds? I would show them.

File: src/main/webapp/app/quiz/participate/quiz.controller.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            // update submission time
+            if (vm.submission && vm.submission.adjustedSubmissionDate) {
+                vm.lastSubmissionTimeText = moment(vm.submission.adjustedSubmissionDate).fromNow();
+            }
+        }, 100);
+
+        vm.onSubmit = onSubmit;
+
+        load();
+
```
Comment (krusche) [inline src/main/webapp/app/quiz/participate/quiz.controller.js]: Please add better comments / documentation

File: src/main/webapp/i18n/de/global.json [hunk 12]
```diff
@@ -116,7 +116,8 @@
             "view": "Details",
             "preview": "Vorschau",
             "cleanup": "Aufräumen",
-            "archive": "Archivieren"
+            "archive": "Archivieren",
+            "submit": "Submit"
```
Comment (krusche) [inline src/main/webapp/i18n/de/global.json]: 'submit' should be 'Absenden' oder 'Einreichen' in German

File: src/main/webapp/i18n/de/question.json [hunk 13]
```diff
@@ -19,7 +19,8 @@
             "text": "Text",
```
Comment (krusche) [inline src/main/webapp/i18n/de/question.json]: Please translate everything to German ;-)

File: src/main/webapp/i18n/de/quizExercise.json [hunk 14]
```diff
@@ -37,7 +37,10 @@
             "secondsShort": "s",
             "questions": "Questions",
             "status": "Status",
-            "startTime": "Start Time"
```
Comment (krusche) [inline src/main/webapp/i18n/de/quizExercise.json]: Please translate everything to German ;-)

File: src/main/webapp/i18n/de/submittedAnswer.json [hunk 15]
```diff
@@ -15,6 +15,7 @@
             "detail": {
```
Comment (krusche) [inline src/main/webapp/i18n/de/submittedAnswer.json]: Please translate everything to German ;-)

**Review by krusche — CHANGES_REQUESTED** [review]
Good progress! Please stick to general coding guidelines and document / comment the code properly.

File: src/main/webapp/app/courses/exercises/exercise-list.html (hunk 6 above)
Comment (VSchlattinger) [reply src/main/webapp/app/courses/exercises/exercise-list.html]: I would either keep "Resume Exercise" or also change the "Start Exercise" to "Start Quiz" (if the exercise is a quiz) to be consistent.

Comment (krusche) [reply src/main/webapp/app/courses/exercises/exercise-list.html]: Then please change it to "Start Quiz" as well

**Review by krusche — APPROVED** [review]
